File: FindMinimumInSortedArray.py
# The approach uses binary search to achieve O(log n) complexity. 
# It checks if the array or subarray is already sorted, in which case the first element is the minimum. 
# Otherwise, it adjusts the search bounds by comparing the middle element with the last element.

class Solution:
    def findMin(self, nums: List[int]) -> int:
        low = 0
        high = len(nums) - 1
        
        while low < high:
            mid = low + (high - low) // 2
            
            if nums[low] < nums[high]:
                return nums[low]
            
            if nums[mid] > nums[high]:
                low = mid + 1
            else:
                high = mid
        
        return nums[low]

File: FindMinimumInSortedArray.java
//The idea here is the the minimum number should be smaller that it's adjacent members. we also added few conditions if the minimum number is at the start or end of the array.
//Time complexity: O(log n)
//Space complexity: O(1)
class Solution {
    public int findMin(int[] nums) {
        int low = 0;
        int high = nums.length-1;
        while(low<=high){
            if(nums[low]<=nums[high]){
            return nums[low];
        }
            int mid = low + (high-low)/2;
            if((mid == 0 || nums[mid]<nums[mid-1]) && (mid == nums.length-1 || nums[mid]<nums[mid+1])){
                return nums[mid];
            }
            else if(nums[low]<=nums[mid]){
                low = mid+1;
            }
            else{
                high = mid-1;
            }
        }
        return -1;
    }
}
